File: PeakElement.java
// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : understanding the logic why BS can be used


// Your code here along with comments explaining your approach in three sentences only
 /*We start by setting low and high pointers and iterates until they converge.
 At each iteration, it compares the current element with its neighbors to determine whether it is a peak, adjusting the pointers accordingly based on the comparison result.
 If a peak is found, its index is returned; otherwise, -1 is returned if no peak exists in the array.
*/
public class PeakElement {
    public int findPeakElement(int[] nums) {
        if (nums.length == 1) return 0;
        int lo = 0, hi = nums.length - 1;
        while (lo <= hi) {
            int mid = lo + (hi - lo)/2;

            if ((mid == nums.length - 1 || nums[mid] > nums[mid + 1]) && (mid == 0 || nums[mid] > nums[mid - 1])) {
                return mid;
            } else if (mid != nums.length - 1 && nums[mid] < nums[mid + 1]) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        return -1;
    }
}

File: peakElement.js
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// Optimized approach
const findPeakElement = (nums) => {
  if (nums.length === 1) {
    return 0;
  }

  if (nums[0] > nums[1]) {
    return 0;
  }

  if (nums[nums.length - 1] > nums[nums.length - 2]) {
    return nums.length - 1;
  }

  let low = 0;
  let high = nums.length - 1;
  let mid;

  while (low <= high) {
    mid = Math.floor((low + high) / 2);
    if (nums[mid] > nums[mid - 1] && nums[mid] > nums[mid + 1]) {
      return mid;
    }

    if (nums[mid + 1] > nums[mid]) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
};

console.log(findPeakElement([1, 2, 1, 3, 5, 6, 4]));

// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

// Brute force approach:
const peakElement = (arr) => {
  console.log("this is arr:", arr);

  if (arr.length === 1) {
    return 0;
  }

  if (arr[0] > arr[1]) {
    return 0;
  }

  if (arr[arr.length - 1] > arr[arr.length - 2]) {
    return arr.length - 1;
  }

  for (let i = 1; i < arr.length - 1; i++) {
    if (arr[i] > arr[i + 1] && arr[i] > arr[i - 1]) {
      return i;
    }
  }
};

console.log(peakElement([1, 2, 3, 1]));
